### ISSUE TITLE
Current Attributes gets reset in tests

### ISSUE THREAD

**Issue by tomprats** [rails/rails#36298]
### Steps to reproduce

- Set adapter to perform jobs inline
  - `ActiveJob::Base.queue_adapter = :inline`
- Set current attribute
  - `Current.foo = "bar"`
- Perform job later
  - `FooJob.perform_later`
- Check value of current attribute

I've created a sample repo that shows the problem with tests. The second commit includes the most relevant code.

https://github.com/tomprats/current-attributes-test

### Expected behavior

- Current attributes should not be nil
  - `Current.foo` => `"bar"`

### Actual behavior

- Current attribute will be null
  - `Current.foo` => `nil`

### System configuration

**Rails version**: 5.2.3

**Ruby version**: 2.5.1

**Comment by fschwahn** [comment]
I can imagine that this is intentional. If you'd change your queue adapter to something "real", `Current` wouldn't be set anymore, leading to bugs right away. If you plan to stay with the `inline`-adapter forever, there's no reason to use activejob.

**Comment by rafaelfranca** [comment]
Yes, this is intentional. The state of current attributes should not be reset after a job is executed. Thank you for the issue.

**Comment by tomprats** [comment]
The problem is that `Current` gets reset for the actual test. I would have thought that the background job would have blank current attributes while the test or request would retain the current attributes.

It seems weird that unrelated code (starting a job) would affect the state of current attributes at all.

I'm also only running into this problem for tests, otherwise I have the adapter set to Sidekiq

**Comment by fschwahn** [comment]
Sorry, I completely misunderstood - I thought you wanted to access `Current ` _in_ the job, not after enqueuing it. Here are the test from your repo, I think these illustrate the problem quite nicely.

```ruby
class FooTest < ActiveSupport::TestCase
  test "Current keeps value" do
    Current.foo = "bar"

    assert_equal Current.foo, "bar"
  end

  test "Current keeps value when job is queued now" do
    Current.foo = "bar"
    FooJob.perform_now

    assert_equal Current.foo, "bar"
  end

  test "Current keeps value when job is queued later" do
    Current.foo = "bar"
    FooJob.perform_later

    assert_equal Current.foo, "bar"
  end
end
```

But it seems @rafaelfranca is saying that this is the expected behavior?

**Comment by kaspth** [comment]
> It seems weird that unrelated code (starting a job) would affect the state of current attributes at all.

It's not that unrelated since the inline adapter runs the jobs in the same process and since jobs are wrapped in the executor the attributes are cleared.

**Comment by tomprats** [comment]
@kaspth That's true, it's related but just wasn't obvious to me at first

Might be good to document it somewhere, but the solution is to use a different queue adapter like https://api.rubyonrails.org/classes/ActiveJob/QueueAdapters/TestAdapter.html

**Comment by allard** [comment]
I've recently been bitten by this, in the case where I ran an ActiveJob as a Callback, and wanted to test 

```Ruby
class Current < ActiveSupport::CurrentAttributes
  attribute :domain # Contains current domain information such as URL
end

class User < ApplicationRecord
  after_create :send_notification

  def send_notification
    NotificationJob.perform_later
  end
end

def SignupController < ApplicationController
  def create
    User.create!(user_params)
    redirect_to some_url(hostname: Current.domain.hostname)
  end
end

class SignupControllerTest < ActionController::TestCase
  test "some test" do
    perform_enqueued_jobs do # I want to test that the notification happen
      post :create, params: {email: "user@company.com" }
    end
  end
end
```

This will fail (in the test) because the after_create callback will cause the job to be executed, which will reset Current and when redirecting in the Controller, the Current.domain is no longer set.

Maybe this could solved by using before/after_reset in Current when running in test, but I'm on Rails 5.2 so don't have before_reset. My solution was to encapsulate the job like:

```ruby
def send_notification
  Current.set(domain: Current.domain) do
    NotificationJob.perform_later
  end
end
```

Posting here in case someone else will find this useful.

I absolutely get why it's important in production to reset Current after each ActiveJob, and also why in testing we want to mimic production as close as we can. But it feels to me that there could be a more elegant solution to this. Maybe ```perform_enqueued_jobs(reset: false)``` or something like that, that would not reset Current in that block, for situations like in this example, where I ended up having to patch my production code in order to make the test run.

**Comment by patodevilla** [comment]
To prevent patching my production code, I did the following on my `Current` class. Not the most elegant solution but its simple and works 😬

```ruby
class Current < ActiveSupport::CurrentAttributes

  if Rails.env.test?
    alias :original_reset :reset

    def reset
      run_callbacks :reset do
        # self.attributes = {}
      end
    end
  end

end
```

If you really need to reset, just call `original_reset`.

**Comment by rience** [comment]
Hey @rafaelfranca,

You're saying that current attribute should not be reset after job is run. And I agree with that.

But then you've closed this issue - and looking at Rails 7.1 this issue is still there. I don't see any commits related to this issue here. 

Has this been fixed and just not linked?

**Comment by bassemawhoob** [comment]
For anyone still looking for a solution, check https://github.com/rails/rails/issues/49227#issuecomment-1900451792
